### PR TITLE
add mesh_vpn data to nodes.json

### PR DESCRIPTION
**Pull request by t-8ch**
Needed by ffnord/meshviewer#10

Example output in statistics object:

```json
{
    "mesh_vpn": {
        "groups": {
            "backbone": {
                "peers": {
                    "w7": null,
                    "w8": null,
                    "w5": null,
                    "w6": {
                        "established": 819191.368
                    },
                    "w3": null,
                    "w0": null,
                    "w9": null,
                    "w4": {
                        "established": 398400.774
                    },
                    "w1": null,
                    "w2": null
                }
            }
        }
    }
}
```

### CODE REVIEW

**Comment by tcatm** [comment]
I think this is way too detailed and hardcoded to fit fastd. Some communities are already using l2tp or other mechanism. Can we reduce this to a flag in "flags"? Maybe call it "uplink"?

**Comment by t-8ch** [comment]
@tcatm Sure, will do.

**Comment by tcatm** [comment]
Great. Does this work for nested groups, too?

**Comment by t-8ch** [comment]
How do they look?

On December 28, 2015 8:09:12 PM GMT+01:00, Nils Schneider <notifications@github.com> wrote:
>Great. Does this work for nested groups, too?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/ffnord/ffmap-backend/pull/76#issuecomment-167630290

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by plumpudding** [comment]
This is always true.

**Comment by tcatm** [comment]
Nested peer groups will look like this:

```
groups { 
   foo {
       peers { ... }
       groups {
           ...
       }
   }
}

File: lib/nodes.py [hunk 1]
```diff
@@ -56,6 +56,23 @@ def mark_online(node, now):
...
 
 
+def mark_uplinks(nodeinfo, node):
+    mark_fastd_uplinks(nodeinfo, node)
+
+
+def mark_fastd_uplinks(nodeinfo, node):
+        try:
+            for group in nodeinfo['statistics']['mesh_vpn']['groups'].values():
+                for peer in group['peer']:
```
Comment (rotanid) [inline lib/nodes.py]: shouldn't it be "in group['peers']" , plural?

File: lib/nodes.py [hunk 2]
```diff
@@ -63,6 +80,8 @@ def import_nodeinfo(nodes, nodeinfos, now, assume_online=False):
         node['flags']['online'] = False
         node['flags']['gateway'] = False
 
+        mark_uplinks(nodeinfo, node)
```
Comment (rotanid) [inline lib/nodes.py]: why isn't mark_fastd_uplinks(nodeinfo, node) called directly?
if there might be other types of uplinks in the future, one could still add the layer.

File: lib/nodes.py [hunk 3]
```diff
@@ -56,6 +56,23 @@ def mark_online(node, now):
     node['flags']['online'] = True
 
 
+def mark_uplinks(nodeinfo, node):
+    mark_fastd_uplinks(nodeinfo, node)
+
+
+def mark_fastd_uplinks(nodeinfo, node):
```
Comment (rotanid) [inline lib/nodes.py]: @tcatm suggested to use singular ( mark_fastd_uplink() ), as there is only one mark per node

**Pull request closed without merging**